### PR TITLE
Support V3 transactions in the `Contract` class

### PR DESCRIPTION
**Pull request by ddoktorski**
Related to #1235 
Initial PR #1255

## Introduced changes

- `PreparedFunctionCall` class:
  - `PreparedFunctionCall` has only methods `call()` and `call_raw()`
  - Add abstract class `PreparedFunctionInvoke` and derive from it `PreparedFunctionInvokeV1` and `PreparedFunctionInvokeV3` with methods `invoke()` and `estimate_fee()`
- `Contract` class:
  - Break down `declare()` into:
    - `declare_v1()`, `declare_v2()` and `declare_v3()`
  - Break down `deploy_contract()` into:
    - `deploy_contract_v1()`, `deploy_contract_v3()` and add `salt` and `unique` parameters
- `ContractFunction` class:
  - Break down `prepare()` into:
    - `prepare_invoke_v1()`, `prepare_invoke_v3()` and `prepare_call()`
  - Break down `invoke()` into:
    - `invoke_v1()` and `invoke_v3()`
- `DeclareResult` class:
  - Break down `deploy()` into:
    - `deploy_v1()` and `deploy_v3()`
- Refactoring:
  - ~Move `SentTransaction` class to `sent_transaction.py`~
  - ~Move `ContractFunction`, `ContractData`, `InvokeResult` and prepared call/invoke related classes to `contract_function.py`~
  - Create `contract_utils.py` file
  - ~Move all `Contract` related classes to the `starknet_py/net/contract` directory~

##

Please note that documentation update and support for `Account.deploy_account()` will be addressed in separate PRs.

- [X] This PR contains breaking changes

### CODE REVIEW

**Comment by ddoktorski** [comment]
Regarding the refactor, splitting the code into files may not look perfect. It's tough to optimally split these closely tied classes without stumbling into circular import issues. If this refactoring doesn't make sense or complicates the review, let me know so I will revert these changes.

File: starknet_py/net/contract/contract_function.py [hunk 1]
```diff
@@ -0,0 +1,481 @@
...
+        :param auto_estimate: Use automatic fee estimation, not recommend as it may lead to high costs.
+        :param nonce: Nonce of the transaction.
+        :return: InvokeResult.
+        """
+        prepared_invoke = self.prepare_invoke_v1(*args, **kwargs)
+        return await prepared_invoke.invoke(
+            max_fee=max_fee, nonce=nonce, auto_estimate=auto_estimate
+        )
+
+    def prepare_invoke_v3(
```
Comment (THenry14) [inline starknet_py/net/contract/contract_function.py:422]: I'm aware of the fact that we agreed to use postfixes. I do wonder, however, **if in this particular case**, where (prepare)_invoke(v1 | v3) can be differed by max_fee/resource_bounds and this is a helper class, would it make sense to have one prepare_invoke and one invoke, that takes those values as Union and based on type it prepares/makes v1 or v3 tx? wdyt? I know it may look like breaking existing approach to interface, but I do wonder if it does make sense in a helper class

Comment (ddoktorski) [reply starknet_py/net/contract/contract_function.py:422]: I definitely don't like the current interface. However, setting a fee as `Union` may not be the best approach:
- A unified name for `max_fee` and `l1_resource_bounds` would be required, let's say just `fee`. While this could make sense for a user with plenty of context, it might be confusing in general
- Thinking long term we will need to incorporate other V3 fields such as `l2_resource_bounds`, `tip` etc. As a result, we would need to alter this interface again

One potential solution is to create `InvokeV1Params` and `InvokeV3Params` and then use 
```
prepare_invoke(params: InvokeParams)
```
However, this approach will break consistency with other parts of our codebase.

Comment (THenry14) [reply starknet_py/net/contract/contract_function.py:422]: Sure, it was just a rough idea either way. We can leave this as is.

**Review by THenry14 — APPROVED** [review]
lgtm

File: starknet_py/net/contract/contract_function.py [hunk 2]
```diff
@@ -0,0 +1,481 @@
...
+    """
+    Basic data of a deployed contract.
+    """
+
+    address: int
+    abi: ABI
+    cairo_version: int
+
+    @cached_property
+    def parsed_abi(self) -> Union[Abi, AbiV1, AbiV2]:
```
Comment (DelevoXDG) [inline starknet_py/net/contract/contract_function.py:44]: Maybe
```suggestion
    def parsed_abi(self) -> Union[AbiV0, AbiV1, AbiV2]:
```
?

File: starknet_py/net/contract/contract_function.py [hunk 3]
```diff
@@ -0,0 +1,481 @@
...
+    invoke_transaction: Invoke = None  # pyright: ignore
+    """A InvokeTransaction instance used."""
+
+    def __post_init__(self):
+        assert self.contract is not None
+        assert self.invoke_transaction is not None
+
+
+@dataclass
+class CallToSend(Call):
```
Comment (DelevoXDG) [inline starknet_py/net/contract/contract_function.py:93]: Have you consider any other names for this one? 😅 

I'm thinking `PreparedCall` or `PreparedCallBase`.

File: starknet_py/net/contract/contract_function.py [hunk 4]
```diff
@@ -0,0 +1,481 @@
...
+        :param block_number: Optional block number.
+        :return: TupleDataclass representing call result.
+        """
+        result = await self.call_raw(block_hash=block_hash, block_number=block_number)
+        return self._payload_transformer.deserialize(result)
+
+
+@add_sync_methods
+@dataclass
+class PreparedFunctionInvoke(ABC, CallToSend):
```
Comment (DelevoXDG) [inline starknet_py/net/contract/contract_function.py:135]: A bit unsure about class name ending with a verb

File: starknet_py/net/contract/contract_function.py [hunk 5]
```diff
@@ -0,0 +1,481 @@
...
+        return await self._invoke(transaction)
+
+    async def estimate_fee(
+        self,
+        block_hash: Optional[Union[Hash, Tag]] = None,
+        block_number: Optional[Union[int, Tag]] = None,
+        *,
+        nonce: Optional[int] = None,
+    ) -> EstimatedFee:
+        tx = await self.get_account.sign_invoke_v1_transaction(
```
Comment (DelevoXDG) [inline starknet_py/net/contract/contract_function.py:224]: I looked into `sign_invoke_v1_transaction`. Why aren't we using query versions for fee estimation?

I understand that it was this way done before, but since RPC 0.6.0 reintroduced query versions, maybe it would make sense to adopt a more correct approach, either now or in a separate PR in the nearby future.

File: starknet_py/net/contract/sent_transaction.py [hunk 6]
```diff
@@ -0,0 +1,57 @@
...
+    ) -> TypeSentTransaction:
+        """
+        Waits for transaction to be accepted on chain till ``ACCEPTED`` status.
+        Returns a new SentTransaction instance, **does not mutate original instance**.
+        """
+        if wait_for_accept is not None:
+            warnings.warn(
+                "Parameter `wait_for_accept` has been deprecated - since Starknet 0.12.0, transactions in a PENDING"
+                " block have status ACCEPTED_ON_L2."
+            )
```
Comment (DelevoXDG) [inline starknet_py/net/contract/sent_transaction.py:43]: Current starknet.py version is not compatible with 0.12.0 anymore, so we can probably remove this, right?

File: starknet_py/tests/e2e/client/full_node_test.py
Comment (DelevoXDG) [inline starknet_py/tests/e2e/client/full_node_test.py:1]: Are you planning to add `invoke_v3` tests in this file as well?

Comment (ddoktorski) [reply starknet_py/tests/e2e/client/full_node_test.py:1]: Thank you for pointing this out. Will add some tests for `invoke_v3` here.

File: starknet_py/net/contract/contract.py
[the remark this answers is not in the document]
Comment (ddoktorski) [reply starknet_py/net/contract/contract.py:1]: This is to improve file organization. I would prefer to move it to `starknet_py/contract`, but considering that we already have `starknet_py/net/account` in place, I think it makes sense to keep contract directory in `starknet_py/net` as well.

File: starknet_py/net/contract/sent_transaction.py (hunk 6 above)
Comment (ddoktorski) [reply starknet_py/net/contract/sent_transaction.py:43]: Will be done in https://github.com/software-mansion/starknet.py/issues/1268, as there is the same parameter to remove in the `Client.wait_for_tx()`.

File: starknet_py/net/contract/contract.py [hunk 7]
```diff
@@ -180,6 +53,9 @@ class DeclareResult(SentTransaction):
     compiled_contract: str = None  # pyright: ignore
     """Compiled contract that was declared."""
 
+    declare_transaction: Declare = None  # pyright: ignore
+    """A Declare transaction that has been sent."""
+
```
[the remark this answers is not in the document]
Comment (ddoktorski) [reply starknet_py/net/contract/contract.py:58]: It is just a helper really. It doesn't require any additional effort from users, and I think it is simply convenient to have. I also think it makes sense that `SentTransaction` contains the actual transaction that has been sent. Regarding the duplicated fields, I agree that there is some space for improvement, but I wouldn't consider this a big concern.

File: starknet_py/net/contract/contract_function.py (hunk 3 above)
Comment (ddoktorski) [reply starknet_py/net/contract/contract_function.py:93]: Will be changed to `PrepareCallBase` 👍

File: starknet_py/net/contract/contract_function.py (hunk 4 above)
Comment (ddoktorski) [reply starknet_py/net/contract/contract_function.py:135]: Agree that this this is not an ideal name. I would rather prefer to avoid `FunctionInvocation`, since it has a [different meaning](https://github.com/starkware-libs/starknet-specs/blob/v0.6.0/api/starknet_trace_api_openrpc.json#L300) in the RPC specification.

File: starknet_py/tests/e2e/client/full_node_test.py
Comment (ddoktorski) [reply starknet_py/tests/e2e/client/full_node_test.py:1]: On second thoughts, these tests are actually meant for `FullNodeClient.get_events()` and `invoke_v1` is used here only as a helper function. We could use `invoke_v3` as well, but let's leave as it is.

File: starknet_py/net/contract/contract_function.py (hunk 5 above)
Comment (ddoktorski) [reply starknet_py/net/contract/contract_function.py]: Good catch! Added `sign_for_fee_estimate()` call to `estimate_fee` methods in `PreparedFunctionInvokeV1` and `PreparedFunctionInvokeV3`.

**Comment by ddoktorski** [comment]
After all, I decided to rollback the refactoring changes. This decision was driven by the user perspective, where it only created needless confusion. For instance, `DeclareResult` was found in the contract module, while `InvokeResult` resided in the contract_function module. I would prefer not to have everything in a single file, but I couldn't find a satisfactory way to logically divide these classes. Other than that, the only changes were made to address @DelevoXDG comments and fix docs build.

File: starknet_py/contract.py [hunk 8]
```diff
@@ -3,14 +3,15 @@
...
 from functools import cached_property
-from typing import Dict, List, Optional, Tuple, TypeVar, Union
+from typing import Dict, List, Optional, TypeVar, Union
 
 from marshmallow import ValidationError
 
-from starknet_py.abi.model import Abi
-from starknet_py.abi.parser import AbiParser
+from starknet_py.abi import Abi as AbiV0
+from starknet_py.abi import AbiParser as AbiV0Parser
```
Comment (DelevoXDG) [inline starknet_py/contract.py:14]: Nit: Maybe move it to `starknet_py.abi.v0.model` for consistency with other abis?
Might be done in a separate PR if that would clutter diff.

File: starknet_py/net/contract/contract_function.py (hunk 4 above)
Comment (DelevoXDG) [reply starknet_py/net/contract/contract_function.py]: Make sense, we have `Invoke` class anyway, so that'll do

File: starknet_py/tests/e2e/client/full_node_test.py
Comment (DelevoXDG) [reply starknet_py/tests/e2e/client/full_node_test.py:1]: Sorry, didn't look into exact purpose of the tests. If all the tests indeed use `invoke_v1` as a helper, I agree it's not necessary

File: starknet_py/contract.py [hunk 9]
```diff
@@ -338,7 +356,70 @@ async def call(
...
+            )
+
+    @property
+    def get_account(self):
+        if self._account is not None:
+            return self._account
+
+        raise ValueError(
+            "The account is not defined. It is not possible to send an invoke transaction."
+        )
```
Comment (DelevoXDG) [inline starknet_py/contract.py:380]: 1. Do we really want to expose account there? It like a bit confusing property to have in `PreparedFunctionInvokeVX`, since it's only used as an internal helper and user would probably be aware which account they used to create prepared invoke anyway.
2. Also, could we maybe make `_account` non-optional? We can check whether it's `None` in `prepare_invoke_vX` before creating instance `PreparedFunctionInvokeVX`.

File: starknet_py/contract.py [hunk 10]
```diff
@@ -382,26 +452,75 @@ async def estimate_fee(
...
+    async def invoke(
+        self,
+        l1_resource_bounds: Optional[ResourceBounds] = None,
+        auto_estimate: bool = False,
+        *,
+        nonce: Optional[int] = None,
+    ) -> InvokeResult:
         """
-        Estimate fee for prepared function call.
+        Send an Invoke transaction version 3 for a prepared data.
```
Comment (DelevoXDG) [inline starknet_py/contract.py:487]: Nit:
```suggestion
        Send an Invoke transaction version 3 for the prepared data.
```

File: starknet_py/contract.py [hunk 11]
```diff
@@ -382,26 +452,75 @@ async def estimate_fee(
...
 
-        :param block_hash: Estimate fee at specific block hash.
-        :param block_number: Estimate fee at given block number
-            (or "latest" / "pending" for the latest / pending block), default is "pending".
+        :param l1_resource_bounds: Max amount and max price per unit of L1 gas (in Fri) used when executing
+            this transaction.
+        :param auto_estimate: Use automatic fee estimation, not recommend as it may lead to high costs.
```
Comment (DelevoXDG) [inline starknet_py/contract.py:491]:
```suggestion
        :param auto_estimate: Use automatic fee estimation; not recommended as it may lead to high costs.
```
or 
```suggestion
        :param auto_estimate: Use automatic fee estimation (not recommended, as it may lead to high costs).
```
(both can be found in the codebase)

File: starknet_py/contract.py [hunk 12]
```diff
@@ -701,35 +942,112 @@ async def deploy_contract(
...
         :param account: BaseAccount used to sign and send deploy transaction.
         :param class_hash: The class_hash of the contract to be deployed.
         :param abi: An abi of the contract to be deployed.
         :param constructor_args: a ``list`` or ``dict`` of arguments for the constructor.
         :param deployer_address: Address of the UDC. Is set to the address of
-            the default UDC (same address on mainnet/testnet/devnet) by default.
+            the default UDC (same address on mainnet/goerli/sepolia) by default.
             Must be set when using custom network other than ones listed above.
         :param cairo_version: Version of the Cairo in which contract is written.
+            By default, it is set to 0.
```
Comment (DelevoXDG) [inline starknet_py/contract.py:959]: Nit: I would consider removing default there, as we're slowly moving away from cairo 0.

Alternatively, and that's just a rough idea for the future, we could check the version of declare transaction used to declare contract at given class hash and determine cairo version that way.

File: starknet_py/tests/e2e/docs/code_examples/test_contract_function.py [hunk 13]
```diff
@@ -2,12 +2,23 @@
...
 
 from starknet_py.contract import Contract, ContractFunction
+from starknet_py.net.client_models import ResourceBounds
 
 
-def test_prepare(map_contract: Contract):
-    # docs-start: prepare
-    prepared_function_call = map_contract.functions["put"].prepare(key=10, value=20)
-    # docs-end: prepare
+def test_prepare_invoke_v1(map_contract: Contract):
```
Comment (DelevoXDG) [inline starknet_py/tests/e2e/docs/code_examples/test_contract_function.py:8]: Should we also add `test_prepare_call`?

File: starknet_py/contract.py (hunk 9 above)
Comment (ddoktorski) [reply starknet_py/contract.py:380]: 1. The main motivation here is to avoid pyright errors when using `_account`. We are sure that it isn't `None`, as we make this verification in `__post_init__` (excluding cases where a user might change it manually on an existing `PreparedFunctionInvokeVX` instance). Considering that an account is a prerequisite for transaction invocation, it kind of makes sense to have such a property in `PreparedFunctionInvoke`. As an alternative, I could make `_get_account` internal.
2. The `account` in the `Contract` class might be `None`, resulting in a None when calling prepare_invoke_vX. We can potentially make this assertion there, but then we would need to add this in four places instead of just one.

File: starknet_py/tests/e2e/docs/code_examples/test_contract_function.py (hunk 13 above)
Comment (ddoktorski) [reply starknet_py/tests/e2e/docs/code_examples/test_contract_function.py:8]: I intentionally left out `test_prepare_call` because I think `prepare_call` isn't generally very useful. However I didn't want to completely remove it from the codebase. There are examples for `ContractFunction.call` and `ContractFunction.call_raw`, which should be sufficient.

File: starknet_py/contract.py (hunk 9 above)
Comment (DelevoXDG) [reply starknet_py/contract.py:380]: > As an alternative, I could make _get_account internal.

That is one solution, because it's overly clear what it does in the context of `PreparedFunctionInvokeVX`

> We can potentially make this assertion there, but then we would need to add this in four places instead of just one.

More like 4 instead of 2. Also, can we maybe check `_account` right away when it's passed to constructor?

File: starknet_py/contract.py (hunk 12 above)
Comment (DelevoXDG) [reply starknet_py/contract.py:959]: Are you planning to add an issue for this?

File: starknet_py/tests/e2e/docs/code_examples/test_contract_function.py (hunk 13 above)
Comment (DelevoXDG) [reply starknet_py/tests/e2e/docs/code_examples/test_contract_function.py:8]: Makes sense

**Review by DelevoXDG — APPROVED** [review]
lgtm 🔥

File: starknet_py/contract.py (hunk 8 above)
Comment (ddoktorski) [reply starknet_py/contract.py:14]: Thanks! Will be done in #1270

File: starknet_py/contract.py (hunk 12 above)
Comment (ddoktorski) [reply starknet_py/contract.py:959]: https://github.com/software-mansion/starknet.py/issues/1271

File: starknet_py/contract.py (hunk 9 above)
Comment (ddoktorski) [reply starknet_py/contract.py:380]: > More like 4 instead of 2

It is implemented now only within the abstract `PreparedFunctionInvoke` class

> can we maybe check _account right away when it's passed to constructor?

This is a `dataclass`, so the only option is `__post_init__`
